defi/single-token-liquidity-pool-2/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+            let lp_resource_address = lp_tokens.resource_address();
+
+            let total_lp  = tokens.amount();
+            let total_liquidity = tokens.amount();
+            
+            let liquidity_pool = Self {
+                pool: Vault::with_bucket(tokens),
+
+                lp_mint_badge: Vault::with_bucket(lp_mint_badge),
+                lp_resource_address,
+
+                total_lp,
+                total_liquidity,
+                lp_divisibility,
+            }
+            .instantiate()
+            .globalize();
+
+            // Return the new LiquidityPool component, as well as the initial supply of LP tokens
+            (liquidity_pool,lp_tokens)
+        }
+
+        /// Adds liquidity to this pool and return the LP tokens representing pool shares
+        /// along with any remainder.
+        pub fn add_liquidity(&mut self,
+             mut tokens: Bucket
+        ) -> (Bucket,Bucket) {
+
+            // Get the resource manager of the lp tokens
+            let lp_resource_manager = borrow_resource_manager!(self.lp_resource_address);
+
+            // Mint LP tokens according to the share the provider is contributing 
+            let mut supply_to_mint = tokens.amount() * (self.total_lp / self.total_liquidity);
+            supply_to_mint = supply_to_mint.round(self.lp_divisibility,RoundingMode::TowardsNearestAndHalfTowardsZero);
+
+            let lp_tokens = self.lp_mint_badge.authorize(|| {
+               return lp_resource_manager.mint(supply_to_mint);
+            });
+
+            // Add tokens to the liquidity pool
+            self.pool.put(tokens.take(tokens.amount()));
+
+            // Return the LP tokens along with any remainder
+            (lp_tokens,tokens)
+        }
+
+        /// Collect fee for liquidity provider. will be added to the pool and LP/Token will be ajusted accordingly
+        pub fn collect_fee(&mut self, 
+            mut tokens: Bucket
+        ) -> Bucket {
+
+            self.pool.put(tokens.take(tokens.amount()));
+
+            let lp_resource_manager = borrow_resource_manager!(self.lp_resource_address);
+            if self.pool.amount() != 0.into() && lp_resource_manager.total_supply() != 0.into() {
+                // self.lp_per_asset_ratio = lp_resource_manager.total_supply() / self.pool.amount();
+                self.total_lp= lp_resource_manager.total_supply();
+                self.total_liquidity = self.pool.amount();
+            }
+
+            tokens
+        }
+
+        /// Removes liquidity from this pool.
+        pub fn remove_liquidity(&mut self,
+             lp_tokens: Bucket
+        ) -> Bucket {
+            assert!(
+                self.lp_resource_address == lp_tokens.resource_address(),
+                "Wrong token type passed in"
+            );
+
+            // Withdraw the correct amounts of tokens A and B from reserves
+            let  mut to_remove = lp_tokens.amount()/(self.total_lp/self.total_liquidity);
```

Comment:
The math here does not seem correct here. Shouldn't this be: `lp_tokens.amount() * self.total_liquidity / self.total_lp`?

Reply:
It's the same result (after testing). but I introduced lp_per_asset_ratio = lp_resource_manager.total_supply() / self.pool.amount() and = 1 if the pool is empty